kvserver: investigate whether we can remove support for inline values from rangefeeds

Currently, no users of rangefeeds should ever see an inline value as they all operate on normal SQL tables.

If we were to run rangefeed on range with inline values, we have at least a few problems:

- Since inline values do not have associated timestamps, we currently emit all inline values encountered during a catchup scan, even if they have not changed since the given cursor.
- If the inline value changed multiple times since the checkpoint, a catchup-scan will only observe the final value.
- If a TBI is used for the catchup scan we may miss the inline value completely.

Do we have future uses for rangefeeds over ranges with inline values?

Jira issue: CRDB-9556

Confirmed, writing an inline value across a rangefeed span panics the node, so better to drop them explicitly:

```
panic: resolved timestamp 1654607236.262852282,0 equal to or above timestamp of operation {key:"\372\022i\000\001" timestamp:<> value:"l>&{\001\022"  <nil> <nil> <nil> <nil> <nil>} [recovered]
```

--- a/pkg/kv/kvclient/kvcoord/dist_sender_rangefeed.go
+++ b/pkg/kv/kvclient/kvcoord/dist_sender_rangefeed.go
@@ -76,6 +76,10 @@ func maxConcurrentCatchupScans(sv *settings.Values) int {
 // Note that the timestamps in RangeFeedCheckpoint events that are streamed back
 // may be lower than the timestamp given here.
 //
+// Rangefeeds do not support inline (unversioned) values, and may omit them or
+// error on them. Similarly, rangefeeds will error if MVCC history is mutated
+// via e.g. ClearRange. Do not use rangefeeds across such key spans.
+//
 // NB: the given startAfter timestamp is exclusive, i.e. the first possible
 // emitted event (including catchup scans) will be at startAfter.Next().
 func (ds *DistSender) RangeFeed(

--- a/pkg/kv/kvclient/rangefeed/rangefeed.go
+++ b/pkg/kv/kvclient/rangefeed/rangefeed.go
@@ -110,10 +110,12 @@ func newFactory(stopper *stop.Stopper, client DB, knobs *TestingKnobs) *Factory 
 // RangeFeed constructs a new rangefeed and runs it in an async task.
 //
 // The rangefeed can be stopped via Close(); otherwise, it will stop when the
-// server shuts down.
+// server shuts down. The only error which can be returned will indicate that
+// the server is being shut down.
 //
-// The only error which can be returned will indicate that the server is being
-// shut down.
+// Rangefeeds do not support inline (unversioned) values, and may omit them or
+// error on them. Similarly, rangefeeds will error if MVCC history is mutated
+// via e.g. ClearRange. Do not use rangefeeds across such key spans.
 //
 // NB: for the rangefeed itself, initialTimestamp is exclusive, i.e. the first
 // possible event emitted by the server (including the catchup scan) is at

--- a/pkg/kv/kvserver/batcheval/cmd_add_sstable_test.go
+++ b/pkg/kv/kvserver/batcheval/cmd_add_sstable_test.go
@@ -760,12 +760,8 @@ func TestEvalAddSSTable(t *testing.T) {
 						}
 
 						// Scan resulting data from engine.
-						iter := storage.NewMVCCIncrementalIterator(engine, storage.MVCCIncrementalIterOptions{
-							EndKey:       keys.MaxKey,
-							StartTime:    hlc.MinTimestamp,
-							EndTime:      hlc.MaxTimestamp,
-							IntentPolicy: storage.MVCCIncrementalIterIntentPolicyEmit,
-							InlinePolicy: storage.MVCCIncrementalIterInlinePolicyEmit,
+						iter := engine.NewMVCCIterator(storage.MVCCKeyAndIntentsIterKind, storage.IterOptions{
+							UpperBound: keys.MaxKey,
 						})
 						defer iter.Close()
 						iter.SeekGE(storage.MVCCKey{Key: keys.SystemPrefix})

--- a/pkg/kv/kvserver/rangefeed/catchup_scan.go
+++ b/pkg/kv/kvserver/rangefeed/catchup_scan.go
@@ -74,12 +74,6 @@ func NewCatchUpIterator(
 				// over the provisional values during
 				// iteration.
 				IntentPolicy: storage.MVCCIncrementalIterIntentPolicyEmit,
-				// CatchUpScan currently emits all inline
-				// values it encounters.
-				//
-				// TODO(ssd): Re-evalutate if this behavior is
-				// still needed (#69357).
-				InlinePolicy: storage.MVCCIncrementalIterInlinePolicyEmit,
 			}),
 		close:     closer,
 		span:      span,
@@ -148,56 +142,48 @@ func (i *CatchUpIterator) CatchUpScan(outputFn outputEventFn, withDiff bool) err
 
 		unsafeKey := i.UnsafeKey()
 		unsafeValRaw := i.UnsafeValue()
-		var unsafeVal []byte
 		if !unsafeKey.IsValue() {
 			// Found a metadata key.
 			if err := protoutil.Unmarshal(unsafeValRaw, &meta); err != nil {
 				return errors.Wrapf(err, "unmarshaling mvcc meta: %v", unsafeKey)
 			}
-			if !meta.IsInline() {
-				// This is an MVCCMetadata key for an intent. The catchUp scan
-				// only cares about committed values, so ignore this and skip past
-				// the corresponding provisional key-value. To do this, iterate to
-				// the provisional key-value, validate its timestamp, then iterate
-				// again. When using MVCCIncrementalIterator we know that the
-				// provisional value will also be within the time bounds so we use
-				// Next.
-				i.Next()
-				if ok, err := i.Valid(); err != nil {
-					return errors.Wrap(err, "iterating to provisional value for intent")
-				} else if !ok {
-					return errors.Errorf("expected provisional value for intent")
-				}
-				if !meta.Timestamp.ToTimestamp().EqOrdering(i.UnsafeKey().Timestamp) {
-					return errors.Errorf("expected provisional value for intent with ts %s, found %s",
-						meta.Timestamp, i.UnsafeKey().Timestamp)
-				}
-				i.Next()
-				continue
+
+			// Inline values are unsupported by rangefeeds. MVCCIncrementalIterator
+			// should have errored on them already.
+			if meta.IsInline() {
+				return errors.AssertionFailedf("unexpected inline key %s", unsafeKey)
 			}
 
-			// If write is inline, it doesn't have a timestamp so we don't
-			// filter on the registration's starting timestamp. Instead, we
-			// return all inline writes.
-			//
-			// TODO(ssd): Do we want to continue to
-			// support inline values here at all? TBI may
-			// miss inline values completely and normal
-			// iterators may result in the rangefeed not
-			// seeing some intermediate values.
-			unsafeVal = meta.RawBytes
-		} else {
-			mvccVal, err := storage.DecodeMVCCValue(unsafeValRaw)
-			if err != nil {
-				return errors.Wrapf(err, "decoding mvcc value: %v", unsafeKey)
+			// This is an MVCCMetadata key for an intent. The catchUp scan only cares
+			// about committed values, so ignore this and skip past the corresponding
+			// provisional key-value. To do this, iterate to the provisional
+			// key-value, validate its timestamp, then iterate again. When using
+			// MVCCIncrementalIterator we know that the provisional value will also be
+			// within the time bounds so we use Next.
+			i.Next()
+			if ok, err := i.Valid(); err != nil {
+				return errors.Wrap(err, "iterating to provisional value for intent")
+			} else if !ok {
+				return errors.Errorf("expected provisional value for intent")
 			}
-			unsafeVal = mvccVal.Value.RawBytes
+			if !meta.Timestamp.ToTimestamp().EqOrdering(i.UnsafeKey().Timestamp) {
+				return errors.Errorf("expected provisional value for intent with ts %s, found %s",
+					meta.Timestamp, i.UnsafeKey().Timestamp)
+			}
+			i.Next()
+			continue
 		}
 
-		// Ignore the version if it's not inline and its timestamp is at
-		// or before the registration's (exclusive) starting timestamp.
+		mvccVal, err := storage.DecodeMVCCValue(unsafeValRaw)
+		if err != nil {
+			return errors.Wrapf(err, "decoding mvcc value: %v", unsafeKey)
+		}
+		unsafeVal := mvccVal.Value.RawBytes
+
+		// Ignore the version if its timestamp is at or before the registration's
+		// (exclusive) starting timestamp.
 		ts := unsafeKey.Timestamp
-		ignore := !(ts.IsEmpty() || i.startTime.Less(ts))
+		ignore := ts.LessEq(i.startTime)
 		if ignore && !withDiff {
 			// Skip all the way to the next key.
 			// NB: fast-path to avoid value copy when !r.withDiff.

--- a/pkg/kv/kvserver/rangefeed/catchup_scan_test.go
+++ b/pkg/kv/kvserver/rangefeed/catchup_scan_test.go
@@ -14,6 +14,7 @@ import (
 	"context"
 	"testing"
 
+	"github.com/cockroachdb/cockroach/pkg/keys"
 	"github.com/cockroachdb/cockroach/pkg/roachpb"
 	"github.com/cockroachdb/cockroach/pkg/storage"
 	"github.com/cockroachdb/cockroach/pkg/storage/enginepb"
@@ -121,4 +122,25 @@ func TestCatchupScan(t *testing.T) {
 		checkEquality(kv2_2_2, kv2_1_1, events[2])
 		checkEquality(kv2_5_3, kv2_2_2, events[3])
 	})
+}
+
+func TestCatchupScanInlineError(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	defer log.Scope(t).Close(t)
+
+	ctx := context.Background()
+	eng := storage.NewDefaultInMemForTesting()
+	defer eng.Close()
+
+	// Write an inline value.
+	require.NoError(t, storage.MVCCPut(ctx, eng, nil, roachpb.Key("inline"), hlc.Timestamp{}, hlc.ClockTimestamp{}, roachpb.MakeValueFromString("foo"), nil))
+
+	// Run a catchup scan across the span and watch it error.
+	span := roachpb.Span{Key: keys.LocalMax, EndKey: keys.MaxKey}
+	iter := NewCatchUpIterator(eng, span, hlc.Timestamp{}, nil)
+	defer iter.Close()
+
+	err := iter.CatchUpScan(nil, false)
+	require.Error(t, err)
+	require.Contains(t, err.Error(), "unexpected inline value")
 }

--- a/pkg/kv/kvserver/rangefeed/processor_test.go
+++ b/pkg/kv/kvserver/rangefeed/processor_test.go
@@ -671,7 +671,6 @@ func TestProcessorInitializeResolvedTimestamp(t *testing.T) {
 	txn1, txn2 := uuid.MakeV4(), uuid.MakeV4()
 	rtsIter := newTestIterator([]storage.MVCCKeyValue{
 		makeKV("a", "val1", 10),
-		makeInline("b", "val2"),
 		makeIntent("c", txn1, "txnKey1", 15),
 		makeProvisionalKV("c", "txnKey1", 15),
 		makeKV("c", "val3", 11),
@@ -680,7 +679,6 @@ func TestProcessorInitializeResolvedTimestamp(t *testing.T) {
 		makeProvisionalKV("d", "txnKey2", 21),
 		makeKV("d", "val5", 20),
 		makeKV("d", "val6", 19),
-		makeInline("g", "val7"),
 		makeKV("m", "val8", 1),
 		makeIntent("n", txn1, "txnKey1", 12),
 		makeProvisionalKV("n", "txnKey1", 12),
@@ -689,7 +687,6 @@ func TestProcessorInitializeResolvedTimestamp(t *testing.T) {
 		makeKV("r", "val9", 4),
 		makeIntent("w", txn1, "txnKey1", 3),
 		makeProvisionalKV("w", "txnKey1", 3),
-		makeInline("x", "val10"),
 		makeIntent("z", txn2, "txnKey2", 21),
 		makeProvisionalKV("z", "txnKey2", 21),
 		makeKV("z", "val11", 4),

--- a/pkg/kv/kvserver/rangefeed/registry_test.go
+++ b/pkg/kv/kvserver/rangefeed/registry_test.go
@@ -169,7 +169,6 @@ func TestRegistrationBasic(t *testing.T) {
 	catchupReg := newTestRegistration(spBC, hlc.Timestamp{WallTime: 1},
 		newTestIterator([]storage.MVCCKeyValue{
 			makeKV("b", "val1", 10),
-			makeInline("ba", "val2"),
 			makeKV("bc", "val3", 11),
 			makeKV("bd", "val4", 9),
 		}, nil), false)
@@ -179,8 +178,8 @@ func TestRegistrationBasic(t *testing.T) {
 	go catchupReg.runOutputLoop(context.Background(), 0)
 	require.NoError(t, catchupReg.waitForCaughtUp())
 	events := catchupReg.stream.Events()
-	require.Equal(t, 6, len(events))
-	require.Equal(t, []*roachpb.RangeFeedEvent{ev1, ev2}, events[4:])
+	require.Equal(t, 5, len(events))
+	require.Equal(t, []*roachpb.RangeFeedEvent{ev1, ev2}, events[3:])
 	catchupReg.disconnect(nil)
 	<-catchupReg.errC
 
@@ -243,7 +242,6 @@ func TestRegistrationCatchUpScan(t *testing.T) {
 	txn1, txn2 := uuid.MakeV4(), uuid.MakeV4()
 	iter := newTestIterator([]storage.MVCCKeyValue{
 		makeKV("a", "valA1", 10),
-		makeInline("b", "valB1"),
 		makeIntent("c", txn1, "txnKeyC", 15),
 		makeProvisionalKV("c", "txnKeyC", 15),
 		makeKV("c", "valC2", 11),
@@ -261,7 +259,6 @@ func TestRegistrationCatchUpScan(t *testing.T) {
 		makeKV("f", "valF3", 7),
 		makeKV("f", "valF2", 6),
 		makeKV("f", "valF1", 5),
-		makeInline("g", "valG1"),
 		makeKV("h", "valH1", 15),
 		makeKV("m", "valM1", 1),
 		makeIntent("n", txn1, "txnKeyN", 12),
@@ -271,7 +268,6 @@ func TestRegistrationCatchUpScan(t *testing.T) {
 		makeKV("r", "valR1", 4),
 		makeIntent("w", txn1, "txnKeyW", 3),
 		makeProvisionalKV("w", "txnKeyW", 3),
-		makeInline("x", "valX1"),
 		makeIntent("z", txn2, "txnKeyZ", 21),
 		makeProvisionalKV("z", "txnKeyZ", 21),
 		makeKV("z", "valZ1", 4),
@@ -326,10 +322,6 @@ func TestRegistrationCatchUpScan(t *testing.T) {
 			roachpb.Key("f"),
 			makeValWithTs("valF3", 7),
 			makeVal("valF2"),
-		),
-		rangeFeedValue(
-			roachpb.Key("g"),
-			makeVal("valG1"),
 		),
 		rangeFeedValue(
 			roachpb.Key("h"),

--- a/pkg/kv/kvserver/rangefeed/task_test.go
+++ b/pkg/kv/kvserver/rangefeed/task_test.go
@@ -64,12 +64,6 @@ func makeMetaKV(key string, meta enginepb.MVCCMetadata) storage.MVCCKeyValue {
 	}
 }
 
-func makeInline(key, val string) storage.MVCCKeyValue {
-	return makeMetaKV(key, enginepb.MVCCMetadata{
-		RawBytes: makeVal(val).RawBytes,
-	})
-}
-
 func makeIntent(key string, txnID uuid.UUID, txnKey string, txnTS int64) storage.MVCCKeyValue {
 	return makeMetaKV(key, enginepb.MVCCMetadata{
 		Txn: &enginepb.TxnMeta{
@@ -254,7 +248,6 @@ func TestInitResolvedTSScan(t *testing.T) {
 		engine := storage.NewDefaultInMemForTesting()
 		testData := []op{
 			{kv: makeKV("a", "val1", 10)},
-			{kv: makeInline("b", "val2")},
 			{kv: makeKV("c", "val4", 9)},
 			{kv: makeKV("c", "val3", 11)},
 			{
@@ -267,7 +260,6 @@ func TestInitResolvedTSScan(t *testing.T) {
 				txn: &txn2,
 				kv:  makeProvisionalKV("d", "txnKey2", 21),
 			},
-			{kv: makeInline("g", "val7")},
 			{kv: makeKV("m", "val8", 1)},
 			{
 				txn: &txn1,
@@ -282,7 +274,6 @@ func TestInitResolvedTSScan(t *testing.T) {
 				txn: &txn1,
 				kv:  makeProvisionalKV("w", "txnKey1", 15),
 			},
-			{kv: makeInline("x", "val10")},
 			{kv: makeKV("z", "val11", 4)},
 			{
 				txn: &txn2,

--- a/pkg/roachpb/api.proto
+++ b/pkg/roachpb/api.proto
@@ -2684,6 +2684,10 @@ message RangeLookupResponse {
 // RangeFeed stream over the provided span, starting at the specified timestamp
 // (exclusive).
 //
+// Rangefeeds do not support inline (unversioned) values, and may omit them or
+// error on them. Similarly, rangefeeds will error if MVCC history is mutated
+// via e.g. ClearRange. Do not use rangefeeds across such key spans.
+//
 // NB: The start timestamp is exclusive, i.e. the first possible event emitted
 // will be at Header.Timestamp.Next(). This includes catchup scans.
 message RangeFeedRequest {

--- a/pkg/storage/enginepb/mvcc3.proto
+++ b/pkg/storage/enginepb/mvcc3.proto
@@ -268,7 +268,7 @@ message RangeAppliedState {
 }
 
 // MVCCWriteValueOp corresponds to a value being written outside of a
-// transaction.
+// transaction. Inline values (without timestamp) are not logged.
 message MVCCWriteValueOp {
   bytes key = 1;
   util.hlc.Timestamp timestamp = 2 [(gogoproto.nullable) = false];

--- a/pkg/storage/mvcc.go
+++ b/pkg/storage/mvcc.go
@@ -1354,12 +1354,6 @@ func mvccPutInternal(
 		if ms != nil {
 			updateStatsForInline(ms, key, origMetaKeySize, origMetaValSize, metaKeySize, metaValSize)
 		}
-		if err == nil {
-			writer.LogLogicalOp(MVCCWriteValueOpType, MVCCLogicalOpDetails{
-				Key:  key,
-				Safe: true,
-			})
-		}
 		return err
 	}
 

--- a/pkg/storage/mvcc_incremental_iterator.go
+++ b/pkg/storage/mvcc_incremental_iterator.go
@@ -24,9 +24,8 @@ import (
 // most recent version (before or at endTime) of that key. If the key was most
 // recently deleted, this is signaled with an empty value.
 //
-// Inline values (non-user data) are handled according to the
-// MVCCIncrementalIterInlinePolicy. By default, an error will be
-// returned.
+// Inline (unversioned) values are not supported, and may return an error or be
+// omitted entirely. The iterator should not be used across such keys.
 //
 // Intents outside the time bounds are ignored. Intents inside the
 // time bounds are handled according to the provided
@@ -87,7 +86,6 @@ type MVCCIncrementalIterator struct {
 
 	// Configuration passed in MVCCIncrementalIterOptions.
 	intentPolicy MVCCIncrementalIterIntentPolicy
-	inlinePolicy MVCCIncrementalIterInlinePolicy
 
 	// Optional collection of intents created on demand when first intent encountered.
 	intents []roachpb.Intent
@@ -123,20 +121,6 @@ const (
 	MVCCIncrementalIterIntentPolicyEmit
 )
 
-// MVCCIncrementalIterInlinePolicy controls how the
-// MVCCIncrementalIterator will handle inline values that it
-// encounters when iterating.
-type MVCCIncrementalIterInlinePolicy int
-
-const (
-	// MVCCIncrementalIterInlinePolicyError will immediately
-	// return an error for any inline value found.
-	MVCCIncrementalIterInlinePolicyError MVCCIncrementalIterInlinePolicy = iota
-	// MVCCIncrementalIterInlinePolicyEmit will return inline
-	// values to the caller.
-	MVCCIncrementalIterInlinePolicyEmit
-)
-
 // MVCCIncrementalIterOptions bundles options for NewMVCCIncrementalIterator.
 type MVCCIncrementalIterOptions struct {
 	EnableTimeBoundIteratorOptimization bool
@@ -147,7 +131,6 @@ type MVCCIncrementalIterOptions struct {
 	EndTime   hlc.Timestamp
 
 	IntentPolicy MVCCIncrementalIterIntentPolicy
-	InlinePolicy MVCCIncrementalIterInlinePolicy
 }
 
 // NewMVCCIncrementalIterator creates an MVCCIncrementalIterator with the
@@ -185,7 +168,6 @@ func NewMVCCIncrementalIterator(
 		endTime:       opts.EndTime,
 		timeBoundIter: timeBoundIter,
 		intentPolicy:  opts.IntentPolicy,
-		inlinePolicy:  opts.InlinePolicy,
 	}
 }
 
@@ -349,16 +331,9 @@ func (i *MVCCIncrementalIterator) initMetaAndCheckForIntentOrInlineError() error
 	}
 
 	if i.meta.IsInline() {
-		switch i.inlinePolicy {
-		case MVCCIncrementalIterInlinePolicyError:
-			i.valid = false
-			i.err = errors.Errorf("unexpected inline value found: %s", unsafeKey.Key)
-			return i.err
-		case MVCCIncrementalIterInlinePolicyEmit:
-			return nil
-		default:
-			return errors.AssertionFailedf("unknown inline policy: %d", i.inlinePolicy)
-		}
+		i.valid = false
+		i.err = errors.Errorf("unexpected inline value found: %s", unsafeKey.Key)
+		return i.err
 	}
 
 	if i.meta.Txn == nil {
@@ -396,13 +371,9 @@ func (i *MVCCIncrementalIterator) initMetaAndCheckForIntentOrInlineError() error
 // advance advances the main iterator until it is referencing a key within
 // (start_time, end_time].
 //
-// It populates i.err with an error if either of the following was encountered:
-//
-// a) an inline value when the inline policy is
-//    MVCCIncrementalIterInlinePolicyError; or
-//
-// b) an intent with a timestamp within the incremental iterator's bounds when
-//    the intent policy is MVCCIncrementalIterIntentPolicyError.
+// It populates i.err with an error if it encountered an inline value or an
+// intent with a timestamp within the incremental iterator's bounds when the
+// intent policy is MVCCIncrementalIterIntentPolicyError.
 func (i *MVCCIncrementalIterator) advance() {
 	for {
 		i.maybeSkipKeys()
@@ -411,14 +382,6 @@ func (i *MVCCIncrementalIterator) advance() {
 		}
 
 		if err := i.initMetaAndCheckForIntentOrInlineError(); err != nil {
-			return
-		}
-
-		// If we have an inline value and the policy was to error, we
-		// would have errored in the call above. If our policy is to
-		// emit inline values, we don't want to advance past it. Inline
-		// values don't have timestamps that we can filter on.
-		if i.meta.IsInline() && i.inlinePolicy == MVCCIncrementalIterInlinePolicyEmit {
 			return
 		}
 
@@ -511,8 +474,8 @@ func (i *MVCCIncrementalIterator) UnsafeValue() []byte {
 
 // NextIgnoringTime returns the next key/value that would be encountered in a
 // non-incremental iteration by moving the underlying non-TBI iterator forward.
-// Intents in the time range (startTime,EndTime] and inline values are handled
-// according to the iterator policy.
+// Intents in the time range (startTime,EndTime] are handled according to the
+// iterator policy.
 func (i *MVCCIncrementalIterator) NextIgnoringTime() {
 	for {
 		i.iter.Next()
@@ -531,15 +494,15 @@ func (i *MVCCIncrementalIterator) NextIgnoringTime() {
 			continue
 		}
 
-		// We have a valid KV or an intent or an inline value to emit.
+		// We have a valid KV or an intent to emit.
 		return
 	}
 }
 
 // NextKeyIgnoringTime returns the next distinct key that would be encountered
 // in a non-incremental iteration by moving the underlying non-TBI iterator
-// forward. Intents in the time range (startTime,EndTime] and inline values are
-// handled according to the iterator policy.
+// forward. Intents in the time range (startTime,EndTime] are handled according
+// to the iterator policy.
 //
 // TODO(sumeer): consider removing this method since it is never used, and it
 // isn't clear what purpose it can serve in the future. We have two current
@@ -568,7 +531,7 @@ func (i *MVCCIncrementalIterator) NextKeyIgnoringTime() {
 			continue
 		}
 
-		// We have a valid KV or an intent or an inline value to emit.
+		// We have a valid KV or an intent to emit.
 		return
 	}
 }

--- a/pkg/storage/mvcc_incremental_iterator_test.go
+++ b/pkg/storage/mvcc_incremental_iterator_test.go
@@ -801,49 +801,23 @@ func TestMVCCIncrementalIteratorInlinePolicy(t *testing.T) {
 			}
 		}
 		t.Run(engineImpl.name, func(t *testing.T) {
-			t.Run("PolicyError returns error if inline value is found", func(t *testing.T) {
+			t.Run("returns error if inline value is found", func(t *testing.T) {
 				iter := NewMVCCIncrementalIterator(e, MVCCIncrementalIterOptions{
-					EndKey:       keyMax,
-					StartTime:    tsMin,
-					EndTime:      tsMax,
-					InlinePolicy: MVCCIncrementalIterInlinePolicyError,
+					EndKey:    keyMax,
+					StartTime: tsMin,
+					EndTime:   tsMax,
 				})
 				defer iter.Close()
 				iter.SeekGE(MakeMVCCMetadataKey(testKey1))
 				_, err := iter.Valid()
 				assert.EqualError(t, err, "unexpected inline value found: \"/db1\"")
 			})
-			t.Run("PolicyEmit returns inline values to caller", func(t *testing.T) {
-				iter := NewMVCCIncrementalIterator(e, MVCCIncrementalIterOptions{
-					EndKey:       keyMax,
-					StartTime:    tsMin,
-					EndTime:      tsMax,
-					InlinePolicy: MVCCIncrementalIterInlinePolicyEmit,
-				})
-				defer iter.Close()
-				iter.SeekGE(MakeMVCCMetadataKey(testKey1))
-				expectInlineKeyValue(t, iter, inline1_1_1)
-				iter.Next()
-				expectKeyValue(t, iter, kv2_2_2)
-				iter.Next()
-				expectKeyValue(t, iter, kv2_1_1)
-				iter.Next()
-				expectInlineKeyValue(t, iter, inline3_2_1)
-
-				iter.SeekGE(MakeMVCCMetadataKey(testKey2))
-				expectKeyValue(t, iter, kv2_2_2)
-				iter.NextIgnoringTime()
-				expectKeyValue(t, iter, kv2_1_1)
-				iter.NextIgnoringTime()
-				expectInlineKeyValue(t, iter, inline3_2_1)
-			})
-			t.Run("PolicyError returns error on NextIgnoringTime if inline value is found",
+			t.Run("returns error on NextIgnoringTime if inline value is found",
 				func(t *testing.T) {
 					iter := NewMVCCIncrementalIterator(e, MVCCIncrementalIterOptions{
-						EndKey:       keyMax,
-						StartTime:    tsMin,
-						EndTime:      tsMax,
-						InlinePolicy: MVCCIncrementalIterInlinePolicyError,
+						EndKey:    keyMax,
+						StartTime: tsMin,
+						EndTime:   tsMax,
 					})
 					defer iter.Close()
 					iter.SeekGE(MakeMVCCMetadataKey(testKey2))
@@ -1000,24 +974,6 @@ func expectKeyValue(t *testing.T, iter SimpleMVCCIterator, kv MVCCKeyValue) {
 	assert.Equal(t, kv.Key.Timestamp, unsafeKey.Timestamp)
 	assert.Equal(t, kv.Value, unsafeVal)
 
-}
-
-func expectInlineKeyValue(t *testing.T, iter SimpleMVCCIterator, kv MVCCKeyValue) {
-	valid, err := iter.Valid()
-	assert.True(t, valid)
-	assert.NoError(t, err)
-
-	unsafeKey := iter.UnsafeKey()
-	unsafeVal := iter.UnsafeValue()
-
-	var meta enginepb.MVCCMetadata
-	err = protoutil.Unmarshal(unsafeVal, &meta)
-	require.NoError(t, err)
-
-	assert.True(t, meta.IsInline())
-	assert.False(t, unsafeKey.IsValue())
-	assert.True(t, unsafeKey.Key.Equal(kv.Key.Key))
-	assert.Equal(t, kv.Value, meta.RawBytes)
 }
 
 func expectIntent(t *testing.T, iter SimpleMVCCIterator, intent roachpb.Intent) {

--- a/pkg/storage/mvcc_logical_ops.go
+++ b/pkg/storage/mvcc_logical_ops.go
@@ -19,6 +19,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/storage/enginepb"
 	"github.com/cockroachdb/cockroach/pkg/util/bufalloc"
 	"github.com/cockroachdb/cockroach/pkg/util/hlc"
+	"github.com/cockroachdb/errors"
 )
 
 // MVCCLogicalOpType is an enum with values corresponding to each of the
@@ -95,6 +96,14 @@ func (ol *OpLoggerBatch) logLogicalOp(op MVCCLogicalOpType, details MVCCLogicalO
 
 	switch op {
 	case MVCCWriteValueOpType:
+		// Disallow inline values. Emitting these across rangefeeds doesn't make
+		// sense, since they can't be ordered and won't be handled by time-bound
+		// iterators in catchup scans. We could include them in the log and ignore
+		// them (or error) in rangefeeds, but the cost doesn't seem worth it.
+		if details.Timestamp.IsEmpty() {
+			panic(errors.AssertionFailedf("received inline key %s in MVCC logical op log", details.Key))
+		}
+
 		if !details.Safe {
 			ol.opsAlloc, details.Key = ol.opsAlloc.Copy(details.Key, 0)
 		}

--- a/pkg/storage/mvcc_logical_ops_test.go
+++ b/pkg/storage/mvcc_logical_ops_test.go
@@ -110,6 +110,11 @@ func TestMVCCOpLogWriter(t *testing.T) {
 				t.Fatal(err)
 			}
 
+			// Write an inline value. This should be ignored by the log.
+			if err := MVCCPut(ctx, ol, nil, testKey6, hlc.Timestamp{}, hlc.ClockTimestamp{}, value6, nil); err != nil {
+				t.Fatal(err)
+			}
+
 			// Verify that the recorded logical ops match expectations.
 			makeOp := func(val interface{}) enginepb.MVCCLogicalOp {
 				var op enginepb.MVCCLogicalOp

